### PR TITLE
Define more operations in terms of cata, ana, and hylo.

### PR DESCRIPTION
**Pull request by sellout**
This is especially useful for Mu/Nu, since these operations will now use the specialized ana/cata rather than the naïve implementations.

Also greatly reduces the number of places we use general recursion.

### CODE REVIEW

**Comment by sellout** [comment]
@paulp I’m guessing this should remove ~19 of the 68 warnings from the Recursion wart. It’s a start.

File: core/shared/src/main/scala/matryoshka/Recursive.scala [hunk 1]
```diff
@@ -32,35 +32,27 @@ trait Recursive[T] extends Based[T] {
...
-    g(loop(t).copoint)
-  }
+      : A =
+    cata[W[A]](t)(fwa => k(fwa.map(_.cojoin)).map(g)).copoint
 
-  def gcataM[W[_]: Comonad : Traverse, M[_]: Monad, A]
+  def gcataM[W[_]: Comonad: Traverse, M[_]: Monad, A]
     (t: T)
-    (w: DistributiveLaw[Base, (M ∘ W)#λ], g: GAlgebraM[W, M, Base, A])
+    (w: DistributiveLaw[Base, W], g: GAlgebraM[W, M, Base, A])
```
Comment (paulp) [inline core/shared/src/main/scala/matryoshka/Recursive.scala]: Heh.

File: core/shared/src/main/scala/matryoshka/Corecursive.scala [hunk 2]
```diff
@@ -25,7 +25,7 @@ trait Corecursive[T] extends Based[T] {
   def embed(t: Base[T])(implicit BF: Functor[Base]): T
 
   def ana[A](a: A)(f: Coalgebra[Base, A])(implicit BF: Functor[Base]): T =
-    embed(f(a) ∘ (ana(_)(f)))
+    hylo(a)(embed, f)
```
Comment (paulp) [inline core/shared/src/main/scala/matryoshka/Corecursive.scala]: Such a big improvement this way.

**Comment by paulp** [comment]
Very much an improvement. From my POV the entire endeavor is much clearer when the hylomorphism takes center stage and the rest is understood relative to that (e.g. ana and cata as degenerate hylomorphisms) where possible. So to see the implementations take that approach is excellent.

**Comment by Baccata** [comment]
Wow this makes so much sense ! Really helps wrapping my head around the concepts. Seems like it will make for a much simpler expression of why recursion schemes matter.

**Comment by sellout** [comment]
Yeah, I often explain things like `cata(f) === hylo(f, project)`, and the implementations are actually identical, so I don‘t know why I didn’t actually define it this way before. The Recursion wart really helps highlight the things we want to avoid, though 😆

**Comment by sellout** [comment]
@paulp BTW, there are at least a handful of other cases I’d like to fix in this PR if you don’t mind waiting a bit. The shorter warning list made a few other changes obvious.

**Comment by tscholak** [comment]
just out of curiosity, for `cata(f) === hylo(f, project)`, are there impacts on speed, either during compile or run time?

**Comment by sellout** [comment]
@tscholak I haven’t benchmarked it, the implementation is pretty much identical. The only difference is that `project` is passed as a parameter to each recursive call and I’m not sure what the overhead of that might be. Also, there is the indirection of one more function call – cata -> hylo.

**Comment by sellout** [comment]
@paulp Ok, I think that’s all for the time being.

**Comment by sellout** [comment]
@paulp Ok, merged in master and made a couple more small changes. Should be ready for merge for reals.

**Comment by paulp** [comment]
It would be great to leverage this very library to compare the precise operation counts of various approaches. No reason we should have to speculate about what's happening under the hood.

**Pull request merged**